File: buySellStock.js
/**
 * 122. Best Time to Buy and Sell Stock II
 *
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    //buy or sell on each day
    //return the max profit in total

    //loop through prices
    // if prices[i] < prices[i+1] (profitable)
    //  profit += prices[i+1] - prices[i] //acumulates every profit
    //return profit

    let profit = 0;

    //loop to store only the profitable buy/sell
    for(let i = 0; i <prices.length; i++){
        if(prices[i] < prices[i+1]){
            profit += prices[i+1] - prices[i];
        }
    }

    return profit;

};

File: buySellStock.java
//Leetcode: 714. Best Time to Buy and Sell Stock with Transaction Fee
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int n = prices.length;
        int profit = 0;
        int buyPrice = prices[0];
        
        for (int i = 1; i < n; i++) {
            profit = Math.max(prices[i] - buyPrice - fee, profit);
            buyPrice = Math.min(buyPrice, prices[i] - profit);
        }
        
        return profit;
    }
}